src/input/source/evdev/touchscreen.rs

```diff
@@ -0,0 +1,383 @@
+#[derive(Debug, Clone, Copy, Default)]
+enum Orientation {
+    #[default]
+    Normal,
+    RotateLeft,
+    RotateRight,
+    UpsideDown,
+}
+
+impl From<&str> for Orientation {
+    fn from(value: &str) -> Self {
+        match value {
+            "normal" => Self::Normal,
+            "left" => Self::RotateLeft,
+            "right" => Self::RotateRight,
+            "upsidedown" => Self::UpsideDown,
+            _ => Self::Normal,
+        }
+    }
+}
+
+/// TouchState represents the state of a single touch
+#[derive(Debug, Clone)]
+struct TouchState {
+    is_touching: bool,
+    pressure: f64,
+    x: f64,
+    y: f64,
+}
+
+impl Default for TouchState {
+    fn default() -> Self {
+        Self {
+            is_touching: Default::default(),
+            pressure: 1.0,
+            x: Default::default(),
+            y: Default::default(),
+        }
+    }
+}
+
+impl TouchState {
+    /// Rotates the touch input to the given orientation
+    fn rotate(&self, orientation: Orientation) -> Self {
+        let mut value = self.clone();
+        let (x, y) = match orientation {
+            Orientation::Normal => (self.x, self.y),
+            Orientation::UpsideDown => (1.0 - self.x, 1.0 - self.y),
+            Orientation::RotateLeft => (1.0 - self.y, self.x),
+            Orientation::RotateRight => (self.y, 1.0 - self.x),
+        };
+        value.x = x;
+        value.y = y;
+
+        value
+    }
+
+    /// Convert the touch into an InputPlumber value with the given touch index
+    fn to_value(&self, idx: u8) -> InputValue {
+        InputValue::Touch {
+            index: idx,
+            is_touching: self.is_touching,
+            pressure: Some(self.pressure),
+            x: Some(self.x),
+            y: Some(self.y),
+        }
+    }
+
+    /// Convert the touch into an InputPlumber event with the given touch index
+    fn to_native_event(&self, idx: u8) -> NativeEvent {
+        NativeEvent::new(Capability::Touchscreen(Touch::Motion), self.to_value(idx))
+    }
+}
+
+/// Source device implementation for evdev touchscreens
+/// https://www.kernel.org/doc/Documentation/input/multi-touch-protocol.txt
+pub struct TouchscreenEventDevice {
+    device: Device,
+    orientation: Orientation,
+    axes_info: HashMap<AbsoluteAxisCode, AbsInfo>,
+    touch_state: [TouchState; 10], // NOTE: Max of 10 touch inputs
+    dirty_states: HashSet<usize>,
+    last_touch_idx: usize,
+}
+
+impl TouchscreenEventDevice {
+    /// Create a new Touchscreen source device from the given udev info
+    pub fn new(
+        device_info: UdevDevice,
+        config: Option<TouchscreenConfig>,
+    ) -> Result<Self, Box<dyn Error + Send + Sync>> {
+        let path = device_info.devnode();
+        log::debug!("Opening device at: {}", path);
+        let mut device = Device::open(path.clone())?;
+        device.grab()?;
+
+        // Set the device to do non-blocking reads
+        // TODO: use epoll to wake up when data is available
```

Comment:
Maybe we should start implementing epoll with this device?

Reply:
Using epoll will require a larger refactor that we should consider in a separate PR.